src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
@weswigham Yes, but the compiler is by and large constructed with the assumption that once we compute a symbol's *declared* type, that type can and will be cached and never changes. However, to do what you suggest we'd have to compute a new declared type for the `node` symbol in each iteration because `entity` depends on `node`, which depends on `entity`, etc. We're simply not equipped to do that currently.